src/progress_bar.rs

```diff
@@ -506,10 +531,103 @@ impl WeakProgressBar {
     pub fn upgrade(&self) -> Option<ProgressBar> {
         let state = self.state.upgrade()?;
         let pos = self.pos.upgrade()?;
-        Some(ProgressBar { state, pos })
+        let ticker = self.ticker.upgrade()?;
+        Some(ProgressBar { state, pos, ticker })
+    }
+}
+
+pub(crate) struct Ticker {
+    stopping_flag: Arc<Mutex<bool>>,
+    stopping_flag_set: Arc<Condvar>,
+    join_handle: Option<thread::JoinHandle<()>>,
+}
+
+impl Drop for Ticker {
+    fn drop(&mut self) {
+        self.stop();
+        self.join_handle.take().map(|handle| handle.join());
     }
 }
 
+#[cfg(test)]
+static TICKER_RUNNING: AtomicBool = AtomicBool::new(false);
+
+impl Ticker {
+    pub(crate) fn new(interval: Duration, bar_state: &Arc<Mutex<BarState>>) -> Self {
+        debug_assert!(!interval.is_zero());
+
+        #[allow(clippy::mutex_atomic)]
+        let stopping_flag = Arc::new(Mutex::new(false));
+        let stopping_flag_set = Arc::new(Condvar::new());
+        let control = TickerControl {
+            stopping_flag: stopping_flag.clone(),
+            stopping_flag_set: stopping_flag_set.clone(),
+            state: Arc::downgrade(bar_state),
+        };
+
+        let join_handle = thread::spawn(move || control.run(interval));
+        Self {
+            stopping_flag,
+            stopping_flag_set,
+            join_handle: Some(join_handle),
+        }
+    }
+
+    pub(crate) fn stop(&self) {
+        *self.stopping_flag.lock().unwrap() = true;
+        self.stopping_flag_set.notify_one();
+    }
+}
+
+struct TickerControl {
+    stopping_flag: Arc<Mutex<bool>>,
+    stopping_flag_set: Arc<Condvar>,
+    state: Weak<Mutex<BarState>>,
+}
+
+impl TickerControl {
+    fn run(&self, interval: Duration) {
+        #[cfg(test)]
+        TICKER_RUNNING.store(true, Ordering::SeqCst);
+
+        while let Some(arc) = self.state.upgrade() {
+            let mut state = arc.lock().unwrap();
+            if state.state.is_finished() {
+                break;
+            }
+
+            if state.state.tick != 0 {
+                state.state.tick = state.state.tick.saturating_add(1);
+            }
+
+            state.draw(false, Instant::now()).ok();
+
+            drop(state); // Don't forget to drop the lock before sleeping
+            drop(arc); // Also need to drop Arc otherwise BarState won't be dropped
+
+            // Wait for `interval` but return early if we are notified to stop
+            let result = self
```

Comment:
How about we only bind the timeout result here, since we don't need the `MutexGuard` anyway? So something like `let (_, result) = `?